File: packages/dart/test/integration_workflows_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:catalogmx/catalogmx.dart';
import 'package:test/test.dart';

List<Map<String, dynamic>> _loadValidationVectors(String name) {
  final file = File('../shared-data/tests/$name');
  final data = jsonDecode(file.readAsStringSync()) as List<dynamic>;
  return data.cast<Map<String, dynamic>>();
}

String _firstValidValue(String name) {
  final vectors = _loadValidationVectors(name);
  return vectors.firstWhere((vector) => vector['valid'] == true)['value']
      as String;
}

void main() {
  group('integration workflows', () {
    test('validates shared identifiers and calculates ISR', () {
      final validRfc = _firstValidValue('rfc_validation.json');
      final validCurp = _firstValidValue('curp_validation.json');
      final validClabe = _firstValidValue('clabe_validation.json');
      final validNss = _firstValidValue('nss_validation.json');

      expect(validateRFC(validRfc), isTrue);
      expect(validateCURP(validCurp), isTrue);
      expect(validateCLABE(validClabe), isTrue);
      expect(validateNSS(validNss), isTrue);

      final isr = ISRCalculator.calculateISR(
        15000,
        periodo: ISRPeriod.mensual,
        year: ISRYear.year2026,
      );
      expect(isr.isrFinal, greaterThan(0));
      expect(isr.isrFinal, lessThan(isr.ingresoGravable));
    });

    test('reads mexico catalogs in one flow', () {
      final salarios = SalariosMinimosCatalog.getAll();
      final umas = UMACatalog.getAll();

      expect(salarios, isA<List<Map<String, dynamic>>>());
      expect(umas, isA<List<Map<String, dynamic>>>());

      if (salarios.isNotEmpty) {
        final year = (salarios.first['año'] ?? salarios.first['year']) as int;
        final salario = SalariosMinimosCatalog.getByYear(year);
        expect(salario, isNotNull);
        expect(
          (salario?['resto_pais'] ?? salario?['zona_general'] ?? 0) as num,
          greaterThan(0),
        );
      }

      if (umas.isNotEmpty) {
        final year = (umas.first['año'] ?? umas.first['year']) as int;
        final uma = UMACatalog.getByYear(year);
        expect(uma, isNotNull);
        expect((uma?['valor_diario'] ?? uma?['diario'] ?? 0) as num,
            greaterThan(0));
      }
    });
  });
}

File: packages/typescript/tests/integration-workflows.test.ts
import fs from 'fs';
import path from 'path';

import { ISRCalculator } from '../src/calculators';
import { SalariosMinimos, UMACatalog } from '../src/catalogs';
import { validateClabe, validateCurp, validateNss, validateRfc } from '../src/validators';

type ValidationVector = {
  value: string;
  valid: boolean;
};

function loadValidationVectors(fileName: string): ValidationVector[] {
  const filePath = path.resolve(__dirname, `../../shared-data/tests/${fileName}`);
  return JSON.parse(fs.readFileSync(filePath, 'utf-8')) as ValidationVector[];
}

function firstValidValue(fileName: string): string {
  const vectors = loadValidationVectors(fileName);
  const entry = vectors.find((vector) => vector.valid);
  if (!entry) throw new Error(`No valid vector found in ${fileName}`);
  return entry.value;
}

describe('integration workflows', () => {
  test('validates shared identifiers and calculates ISR', () => {
    const validRfc = firstValidValue('rfc_validation.json');
    const validCurp = firstValidValue('curp_validation.json');
    const validClabe = firstValidValue('clabe_validation.json');
    const validNss = firstValidValue('nss_validation.json');

    expect(validateRfc(validRfc)).toBe(true);
    expect(validateCurp(validCurp)).toBe(true);
    expect(validateClabe(validClabe)).toBe(true);
    expect(validateNss(validNss)).toBe(true);

    const isr = ISRCalculator.calcular(15000, 2026, 'mensual', false);
    expect(isr.isr_a_retener).toBeGreaterThan(0);
    expect(isr.isr_a_retener).toBeLessThan(isr.ingreso_gravable);
  });

  test('reads mexico catalogs in the same workflow', () => {
    const salario2024 = SalariosMinimos.getPorAño(2024);
    const uma2024 = UMACatalog.getPorAño(2024);

    expect(salario2024).toBeDefined();
    expect(uma2024).toBeDefined();
    expect((salario2024?.resto_pais ?? salario2024?.zona_general ?? 0)).toBeGreaterThan(0);
    expect((uma2024?.valor_diario ?? 0)).toBeGreaterThan(0);
  });
});
